File: HW08/hw8/src/HashSet.java

/**
 * HashSet class.
 * @author Cihan
 * @param <E> the type of elements in this collection.
 * There is no duplicate elements in this collection.
 */
public class HashSet<E> implements Set<E> {
	/**
	 * Array of E.
	 */
	private E[] elements;
	/**
	 * Size of array that includes element.
	 */
	private int size=0;
	/**
	 * The load factor is a measure of how full the HashSet is allowed to get before its capacity is automatically increased.
	 */
	private double loadFactor = 0.75;
	
	@SuppressWarnings("unchecked")
	public HashSet() {
		elements = (E[]) new Object[8];
	}
	/**
	 * One parameter constructor.
	 * @param capacity initial capacity
	 */
	@SuppressWarnings("unchecked")
	public HashSet(int capacity) {
		elements = (E[]) new Object[capacity];
	}
	
	@Override
	public MyIterator<E> iterator(){
		
		return  new MyIterator<E>() {
			int index = 0;
			
			@Override
			public boolean hasNext() {
				if(index < size) {
					return true;
				}
				return false;
			}
			
			@Override
			public E next() {
				return elements[index++];
			}
			
			@Override
			public E previous() {
				return elements[index--];
			}
			
			@Override
			public int position() {
				return index;
			}
			
			@Override
			public void remove() {
				HashSet.this.remove(elements[index]);
			}
		};
	}
	
	@Override
	public boolean add(E e) {
		
		MyIterator<E> myiter = this.iterator();
		boolean res = true;
		
		while(myiter.hasNext()) { // Checking next.
			if(myiter.next().equals(e)) { // Control for duplicate elements.
				res = false;
			}
		}
		
		if(res) { // Adding it.
			if(shouldGrowSet(elements)) {	//Checking for size and growing up the set.
				growSet();
			}
			elements[size] = e;
			size++;
		}
		
		return res;
	}
	
	@Override
	public boolean addAll(Collection<E> c) {
		if(c.isEmpty()){
			return false;
		}
		
		MyIterator<E> otherIter = c.iterator();

		boolean res=true;

		while(otherIter.hasNext()){
			if(!(add(otherIter.next()))){
				res=false;
			}
		}

		return res;
		
	}
	
	@Override
	public void clear() {
		size = 0;
	}
	
	@Override
	public boolean contains(E e) {
		MyIterator<E> myiter = this.iterator();
		
		while(myiter.hasNext()) {
			if(myiter.next().equals(e)) {
				return true;
			}
		}
		return false;
	}
	
	@Override
	public boolean containsAll(Collection<E> c) {
		
		if(c.isEmpty()){
			return false;
		}

		MyIterator<E> otherIter = c.iterator();

		boolean res=true;

		while(otherIter.hasNext()){
			if(!(contains(otherIter.next()))){
				res = false;
			}
		}

		return res;
	}
	
	@Override
	public boolean isEmpty() {
		return size==0;
	}
	
	@Override
	public boolean remove(E e) {
		MyIterator<E> removeIt = this.iterator();
		
		boolean res = false;
		
		while(removeIt.hasNext()) {
			if(removeIt.next().equals(e)) {
				res = true;
				break;
			}
		}
		
		if(res) {
			for(int i=removeIt.position();i<size;i++) {
				elements[i-1] = elements[i];
			}
			size--;
		}
		return res;
	}
	
	@Override
	public boolean removeAll(Collection<E> c) {
		if(c.isEmpty()){
			return false;
		}
		boolean res=true;
		
		MyIterator<E> otherIter = c.iterator(); // iterator for incoming collection.

		while(otherIter.hasNext()){
			if(!(remove(otherIter.next()))){
				res=false;
			}
		}
		
		return res;
	}
	
	@Override
	public boolean retainAll(Collection<E> c) {
		MyIterator<E> myIter = iterator();
		HashSet<E> temp = new HashSet<E>();	 // creating a temp hashset.

		while(myIter.hasNext()){
			if( !(c.contains(myIter.next())) ){
				myIter.previous();
				temp.add(elements[myIter.position()]);
				myIter.next();
			}
		}

		if(removeAll(temp)) {
			return true;
		}
		return false;
	}
	
	@Override
	public int size() {
		return size;
	}
	
	/**
	 * Checks if the set needs to grow.
	 * @param elementarray
	 * @return true if it need grow else false.
	 */
	private boolean shouldGrowSet(E[] elementarray) {
		if(size > elementarray.length * loadFactor) {
			return true;
		}
		return false;
	}
	
	/**
	 * Growing up the set.
	 */
	@SuppressWarnings("unchecked")
	private void growSet() {
		
		E[] newElements = (E[]) new Object[2*elements.length];
		
		for(int i = 0 ; i<size ; i++) {
			
			newElements[i] = elements[i];
			
		}
		
		elements = (E[]) new Object[2*elements.length];
		
		for(int i = 0 ; i<size ; i++) {
			
			elements[i] = newElements[i];
			
		}	
	}		
}

File: HW06/hashset.cpp
#include "hashset.h"

namespace MyCollection{

	template<typename E,typename container>
	Iterator<E,container> HashSet<E,container>::iterator()  { 
		Iterator<E,container> iter(_c);
		return iter; 
	}

	template<typename E,typename container>
	// Adds the element in this set.
	bool HashSet<E,container>::add(const E &e)  {

		Iterator<E,container> iter = iterator();

		while(iter.hasNext()){	// Checking next.
			if(iter.Next() == e){	// Control for duplicate elements.
				return false;
			}
		}

		_c.insert(iter.getIteratorBegin(),e);	// Adding it.
		return true;

	}
	template<typename E,typename container>
	// Adding all incoming collection's elements to our collection.If we can add all of them then returns true.Otherwise returns false.
	bool HashSet<E,container>::addAll(Collection<E,container > &c)  {
	
		Iterator<E,container> otherIter = c.iterator();

		bool res=true;

		while(otherIter.hasNext()){
			if(!(add(otherIter.Next()))){
				res=false;
			}
		}

		return res;

	}
	template<typename E,typename container>
	//Clearing the container.
	void HashSet<E,container>::clear()  {
		_c.clear();
	}
	template<typename E,typename container>
	//Returns true if this collection contains the specified element
	bool HashSet<E,container>::contains(const E &e)  {
		Iterator<E,container> iter = iterator();

		while(iter.hasNext()){
			if(iter.Next() == e){
				return true;
			}
		}

		return false;
		
	}
	template<typename E,typename container>
	//Returns true if this collection contains all of the elements in the specified collection.
	bool HashSet<E,container>::containsAll(Collection<E,container > &c)  {

		if(c.isEmpty()){
			return false;
		}

		Iterator<E,container> otherIter = c.iterator();

		bool res=true;

		while(otherIter.hasNext()){
			if(!(contains(otherIter.Next()))){
				res = false;
			}
		}

		return res;
	}
	template<typename E,typename container>
	//Returns true if this collection contains no elements
	bool HashSet<E,container>::isEmpty()  {
		return _c.size() == 0;
	}
	template<typename E,typename container>
	//Removes a single instance of the specified element from this collection.
	void HashSet<E,container>::remove(const E &e)  {
		Iterator<E,container> iter = iterator();

		if(*iter.getIteratorBegin() == e){	// If the first element equals to e then removes that.
			_c.erase(iter.getIteratorBegin());
			return;
		}
		else{
			while(iter.hasNext()){
				if(iter.Next() == e){	
					_c.erase(--iter.getIteratorBegin());
					return;
				}
			}
		}

	}
	template<typename E,typename container>
	//Removes all of this collection's elements that are also containedin the specified collection
	void HashSet<E,container>::removeAll(Collection<E,container > &c)  {

		Iterator<E,container> otherIter = c.iterator(); // iterator for incoming collection.

		while(otherIter.hasNext()){
			remove(otherIter.Next());
		}
		
	}
	template<typename E,typename container>
	//Retains only the elements in this collection that are contained in the specified collection
	void HashSet<E,container>::retainAll(Collection<E,container > &c)  {
		Iterator<E,container> myIter = iterator();
		HashSet<E,container > temp;	 // creating a temp hashset.

		while(myIter.hasNext()){
			if( !(c.contains(myIter.Next())) ){
				--myIter;
				temp.add((*myIter.getIteratorBegin()));
				myIter.Next();
			}
		}

		removeAll(temp);

	}
	template<typename E,typename container>
	//Returns the size.
	int HashSet<E,container>::size() {
		return _c.size();
	}

}